src/Lean/Meta/Match/MatcherApp/Transform.lean

```diff
@@ -156,4 +156,219 @@ def refineThrough? (matcherApp : MatcherApp) (e : Expr) :
   catch _ =>
     return none
 
+
+/--
+Given `n` and a type `α₁ → α₂ → ... → αₙ → Sort u`, returns the types `α₁, α₂, ..., αₙ`.
+
+This can be used to infer the expected type of the alternatives when constructing a `MatcherApp`.
+-/
+-- TODO: Which is the natural module for this?
+def arrowDomainsN (n : Nat) (type : Expr) : MetaM (Array Expr) := do
+  let mut type := type
+  let mut ts := #[]
+  for i in [:n] do
+    type ← whnfForall type
+    let Expr.forallE _ α β _ ← pure type | throwError "expected {n} arguments, got {i}"
+    if β.hasLooseBVars then throwError "unexpected dependent type"
+    ts := ts.push α
+    type := β
+  return ts
+
+/--
+Performs a possibly type-changing transformation to a `MatcherApp`.
+
+* `onParams` is run on each parameter and discrimitant
```

Comment:
I don't know the term "discrimitant", and my dictionary doesn't either. Is this specialized technical vocabulary, or should it be "scrutinee"?

Reply:
Scrurtinee seems to be mostly GHC lingo, and `discrs` is used in the lean code base. 🤷‍♂️

Reply:
```suggestion
* `onParams` is run on each parameter and discriminant
```

Reply:
Consistency is good here!

"Scrutinee" is also what we called it in Idris 1, but there were plenty of GHC-isms in that code as well, like "zonk".

Reply:
Though "-ant" is ["a person or thing performing or causing the stated action"](https://dictionary.cambridge.org/dictionary/english/ant?q=-ant). But the thing we're talking about here is not that which is causing discrimination to occur, it's that which is being subject to discrimination. "Discriminee" isn't in any English dictionaries that I could find, but it follows the regular pattern of "-ee" pointing at the object of a verb.

Reply:
Indeed googling shows up Coq docs that (occasionally? rarely?) use of `discriminee`, but never discriminant.

But yet:
```
~/build/lean/lean4 $ git grep discriminant|wc -l
84
~/build/lean/lean4 $ git grep discriminee|wc -l
0
```

I don’t mind changing that, but if we do, then everywhere separate of this PR.

Do we need a little “glossary and nomenclature” document somewhere where we can keep track of such decisions?

Reply:
"scrutinee" is also used in the Rust reference: https://doc.rust-lang.org/reference/expressions/match-expr.html#match-expressions